xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/XWiki.java

```diff
@@ -4354,8 +4370,11 @@ public boolean checkAccess(String action, XWikiDocument doc, XWikiContext contex
         if (action.equals("skin") && SKIN_RESOURCE_SPACE_NAMES.contains(firstSpaceName)) {
             // We still need to call checkAuth to set the proper user.
             XWikiUser user = checkAuth(context);
+
             if (user != null) {
                 context.setUser(user.getUser());
+                UserReference userReference = getUserReferenceResolver().resolve(user.getUserReference());
+                getObservationManager().notify(new UserAuthenticationEvent(userReference), user);
```

Comment:
Note that here you're still sending the event with an XWikiUser as source, contrary to the javadoc of the event. You don't need to send anything else except the UserReference that you put in parameter as Thomas mentioned. So you can probably just send the event (constructed with the UserReference) without more info.

Reply:
Okay, I'll fix that

Reply:
Note that there might be other locations where #checkAuth is called. It would be better to place yourself in XWiki#checkAuth instead of XWiki#checkAccess IMO. Also, I feel you should compare the context user before and after the call to `getAuthService().checkAuth(context)` to send an event only when the user was not already authenticated.

Reply:
> I feel you should compare the context user before and after the call to getAuthService().checkAuth(context) to send an event only when the user was not already authenticated.

Actually it's not going to be enough since this will still produce a new event for every single HTTP request. Right now only the XWikiAuthService knows if it's the same user or a new authentication. @surli where is the authentication protection catch failing auth ?

Reply:
> @surli where is the authentication protection catching failing auth ?

It's performed in various places: the authenticators (MyFormAuthenticator and MyBasicAuthenticator) and also in XWikiSecretVerifier. There was no single point for this AFAIR.

Reply:
Indeed there is also https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-server/src/main/java/org/xwiki/rest/internal/XWikiSecretVerifier.java#L85. I though the REST service was going through MyBasicAuthenticator since it is a BASIC auth in practice.

Reply:
Do I need to setup a whole AuthenticationFailure-like management to achieve the UserAuthenticated thing ? Moreover, the UserAuthenticatedEvent could be a AuthenticationSuccessEvent, couldn't it be ?

Reply:
> Do I need to setup a whole AuthenticationFailure-like management to achieve the UserAuthenticated thing ?

Not sure what you mean: here you only need to trigger the event, but yes you might need to do it in the various location pointed by Thomas. @tmortagne if I'm correct it means we won't get the event for specific authenticators, which doesn't feel right in that case: are you sure it's not possible to just trigger the event where it was originally here?

> Moreover, the UserAuthenticatedEvent could be a AuthenticationSuccessEvent, couldn't it be ?

I'm mixed. I agree that if both events (AuthenticationFailureEvent and the new one) are triggered from same places they should have the same name pattern. Now both events have different usages and are not constructed the same way, so not entirely sure.

Reply:
> @tmortagne if I'm correct it means we won't get the event for specific authenticators, which doesn't feel right in that case: are you sure it's not possible to just trigger the event where it was originally here?

Not really sure what you are proposing. I explained what was the problem with the way it was done originally.